onnxruntime/core/framework/bfc_arena.cc

```diff
@@ -407,6 +418,85 @@ void BFCArena::Free(void* p) {
   }
 }
 
+Status BFCArena::OnRunEnd() {
+  if (shrink_on_every_run_) {
+    return Shrink();
+  }
+  return Status::OK();
+}
+
+Status BFCArena::Shrink() {
+  std::lock_guard<OrtMutex> lock(lock_);
+  auto num_regions = region_manager_.regions().size();
+  std::vector<void*> region_ptrs;
+  std::vector<size_t> region_sizes;
+  region_ptrs.reserve(num_regions);
+  region_sizes.reserve(num_regions);
+  for (const auto& region : region_manager_.regions()) {
+    // Even if any byte is left unused in the first allocation region, we do not want to consider it for de-allocation
+    // We never want to shrink the initial allocation if the arena extend strategy is kNextPowerOfTwo.
+    // This could seem confusingly arbitrary but the rationale is as follows:
+    // The user selected initial allocation chunk is only valid for the arena extend strategy kNextPowerOfTwo
+    // and the user has likely chosen this initial value so that any ad-hoc arena extensions/shrinkages could potentially
+    // be avoided. So we do not consider the initial allocation for shrinkage whatever its usage status.
+    // On the other hand, if the arena extension strategy is kSameAsRequested, any initial chunk set by the user or otherwise,
+    // is moot and the arena will only extend based on the request size. In these cases, we consider any allocation for shrinkage
+    // if it is left unused (even if it is the first allocation).
+    if (arena_extend_strategy_ == ArenaExtendStrategy::kSameAsRequested || region.id() != 0) {
```

Comment:
nit: arena_extend_strategy_ doesn't change for the arena. Do we still need to check each time?

Reply:
refactored code a little bit to avoid multiple checks